core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp

```diff
@@ -0,0 +1,570 @@
+  using work_tag      = typename Policy::work_tag;
+  using launch_bounds = typename Policy::launch_bounds;
+
+  FunctorType const m_functor;
+  Policy const m_policy;
+  size_type const m_league_size;
+  int m_team_size;
+  size_type const m_vector_size;
+  int m_shmem_begin;
+  int m_shmem_size;
+  void* m_scratch_ptr[2];
+  int m_scratch_size[2];
+
+  template <typename Functor>
+  void sycl_direct_launch(const Policy& policy, const Functor& functor) const {
+    // Convenience references
+    const Kokkos::Experimental::SYCL& space = policy.space();
+    Kokkos::Experimental::Impl::SYCLInternal& instance =
+        *space.impl_internal_space_instance();
+    sycl::queue& q = *instance.m_queue;
+
+    q.submit([&](sycl::handler& cgh) {
+      const auto shmem_begin  = m_shmem_begin;
+      const auto shmem_size   = m_shmem_size;
+      const auto scratch_size = m_scratch_size[1];
+      auto team_lambda        = [=](sycl::group<1> g) {
+        // FIXME_SYCL Add scratch memory
+        const member_type team_member(nullptr, shmem_begin, shmem_size, nullptr,
+                                      scratch_size, g);
+
+        if constexpr (std::is_same<work_tag, void>::value)
+          functor(team_member);
+        else
+          functor(work_tag(), team_member);
+      };
+      if (m_team_size > 0)
+        cgh.parallel_for_work_group(sycl::range<1>(m_league_size),
+                                    sycl::range<1>(m_team_size), team_lambda);
+      else
+        cgh.parallel_for_work_group(sycl::range<1>(m_league_size), team_lambda);
+    });
+    space.fence();
+  }
+
+  // Indirectly launch a functor by explicitly creating it in USM shared memory
+  void sycl_indirect_launch() const {
+    // Convenience references
+    const Kokkos::Experimental::SYCL& space = m_policy.space();
+    Kokkos::Experimental::Impl::SYCLInternal& instance =
+        *space.impl_internal_space_instance();
+    Kokkos::Experimental::Impl::SYCLInternal::IndirectKernelMemory& kernelMem =
+        *instance.m_indirectKernel;
+
+    // Allocate USM shared memory for the functor
+    kernelMem.resize(std::max(kernelMem.size(), sizeof(m_functor)));
+
+    // Placement new a copy of functor into USM shared memory
+    //
+    // Store it in a unique_ptr to call its destructor on scope exit
+    std::unique_ptr<FunctorType, Kokkos::Impl::destruct_delete>
+        kernelFunctorPtr(new (kernelMem.data()) FunctorType(m_functor));
+
+    // Use reference_wrapper (because it is both trivially copyable and
+    // invocable) and launch it
+    sycl_direct_launch(m_policy, std::reference_wrapper(*kernelFunctorPtr));
+  }
+
+ public:
+  inline void execute() const {
+    if (m_league_size == 0) return;
+
+    // if the functor is trivially copyable, we can launch it directly;
+    // otherwise, we will launch it indirectly via explicitly creating
+    // it in USM shared memory.
+    if constexpr (std::is_trivially_copyable_v<decltype(m_functor)>)
```

Comment:
Given that `sycl::is_memcpyable` is in SYCL 2020 (although not yet implemented anywhere), should we have a private trait of our own which will currently default to `is_trivially_copyable`, because this check is starting to be in a lot of places.

Reply:
I don't feel strongly either way. It looks like an easy change, though.

Reply:
I don't feel strongly either.  Maybe just mark it as a FIXME_SYCL?

Reply:
Since this is not related to this pull request in particular, just open an issue for it.